### DISCUSSION TITLE
getting_started/first_3d_game/04.mob_scene

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#118] [User-contributed notes]
# getting_started/first_3d_game/04.mob_scene

In this part, you're going to code the monsters, which we'll call mobs. In the next lesson, we'll spawn them randomly around the playable area. Let's design the monsters themselves in a new scene. ...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/04.mob_scene.html

**Comment by vmorris** [comment]
It would be nice if you explained what the local transformation means...

**Reply by Calinou** [reply]
I can't find any references to `local` or `transform` on that page. Which paragraph are you referring to?

**Reply by asperkraken** [reply]
Are you referring to the `look_at_from_position()' arguments? The IDE/editor has a tooltip. But for clarity:

The first argument in look_at_from_position is current position
The second argument is the target position
The third argument determines which direction is "up".
There is also an optional fourth argument which is not needed here.

**Reply by HamzaMateen** [reply]
Local Transformation when applied to an object means that the transformation is taking place w.r.t to the same coordinate system in which the object is placed. 

Transformations such as Scaling and Translation are a couple examples of this. 

By contrast, World Transformation refers to a transformation where the underlying coordinate system itself is changed, e.g. Projection of 3D objects onto a 2D plane. Most accessible example of this is Rendering, where a 3D game object is projected onto the 2D Camera of the game. 

i hope this helps!

**Comment by kuanslove** [comment]
Jus t quick question: 
in the code, 

```gdscript
# This function will be called from the Main scene.
func initialize(start_position, player_position):
	look_at_from_position(start_position, player_position, Vector3.UP)
	rotate_y(randf_range(-PI / 4, PI / 4))
	var random_speed = randi_range(min_speed, max_speed)
	velocity = Vector3.FORWARD * random_speed
	velocity = velocity.rotated(Vector3.UP, rotation.y)
```

After the look_at_from_position, the Mob object already point to the player, and so is the velocity, why do we need to rotate the velocity again with rotation.y at last ?  Could anyone show me with the Vector value change after each step of the code? Or visualize what it looks like


Thanks

**Reply by bavariangolem** [reply]
I can describe it to you qualitativ:

The first part:
```gdscript
# rotates the object
look_at_from_position(start_position, player_position, Vector3.UP)
rotate_y(randf_range(-PI / 4, PI / 4))

# set the velocity from a **global point of view** - it's like "Vector3(0, 0, -1) * random_number"
velocity = Vector3.FORWARD * random_speed

# Rotate the velocity to the direction where the mob is looking at
velocity = velocity.rotated(Vector3.UP, rotation.y)
```

Hope that is helpful

**Reply by RVishnu17** [reply]
Y reply it so complicated ?
```gdscript
func initialize(start_position, player_position):
# basically point towards the player after spawning	
look_at_from_position(start_position, player_position, Vector3.UP)

# we dont want them all coming straight towards us , so they can -45 to 45 fov towards us
rotate_y(randf_range(-PI / 4, PI / 4))

# rest is the velocity for that fov(pi/4 to -pi/4)
	var random_speed = randi_range(min_speed, max_speed)
	velocity = Vector3.FORWARD * random_speed
	velocity = velocity.rotated(Vector3.UP, rotation.y)
```

**Reply by Nedus** [reply]
So lessons learned, _velocity_ is a built in property of _CharacterBody3D_ , it is a Vector3 and it is usually .ZERO. Rotating the mesh has nothing to do with the direction the mesh will move. Also Vector3.FORWARD is a global constant, it is just this: (0, 0, -1), being global means is not a property of the CharacterBody3D. I'm a noob, and this part of the project helped me a lot cause I studied all the concepts one by one, even if not explicitly explained...  it took me a while to get onto it. Isn't all this abstraction and (sort of) decoupling?

**Comment by Justin-Archmagus** [comment]
I would reword the note about memory management and pooling to be less proscribing about its value in GDScript.  This is a good time to make a note about reference counting and how that can make garbage collection behavior different from C++ (hahaha wut) and C#, but that doesn't mean that pooling might not make sense for some games.  There may still be many reasons why a game might want to eagerly allocate and/or keep their objects in memory for reuse besides avoiding a badly timed garbage collection event.

**Reply by wavecommander** [reply]
Yes, exactly. Pooling avoids heap allocations and saves on initialization overhead regardless of garbage collection or reference counting methods.

**Comment by reinhillmann** [comment]
A quick comment about `look_at_from_position`. Later in the course we add the jump ability. If the player is busy jumping, the `look_at_from_position` function will make any mob that spawns during the jump look slightly upwards towards the player. This causes the collision detection algorithm we write later to misbehave.

Instead, the spawning mob should look at the player position only in the `x` and `z` plane. The `y` plane should not be affected.

Change
```gdscript
look_at_from_position(start_position, player_position, Vector3.UP)
```
to
```gdscript
var target = Vector3(player_position.x, start_position.y, player_position.z)
look_at_from_position(start_position, target, Vector3.UP)
```

**Reply by Calinou** [reply]
Pull requests opened:

- https://github.com/godotengine/godot-docs/pull/10666
- https://github.com/godotengine/godot-demo-projects/pull/1165

**Reply by FrackinFamous** [reply]
This may be a bit more concise. 

``` gdscript
look_at_from_position(start_position, player_position, Vector3.UP)
# immediately set the x rotation to zero to stop wheelies. 
rotation.x = 0
```

**Reply by Poobslag** [reply]
I noticed this too -- and more noticably, if the player jumps off the side of the level, all of the mobs will spawn perpendicular to the floor.

**Comment by MetalWhine** [comment]
Hello there, I would like to suggest a small change to this section of the code in C#:

```C#
public void Initialize(Vector3 startPosition, Vector3 playerPosition)
{
    // ...

    // We calculate a random speed (integer).
    int randomSpeed = GD.RandRange(MinSpeed, MaxSpeed);
    // We calculate a forward velocity that represents the speed.
    Velocity = Vector3.Forward * randomSpeed;
    // We then rotate the velocity vector based on the mob's Y rotation
    // in order to move in the direction the mob is looking.
    Velocity = Velocity.Rotated(Vector3.Up, Rotation.Y);
}
```
GD.RandRange() requires an explicit cast so it would be great to add:
```C#
int randomSpeed = (int)GD.RandRange(MinSpeed, MaxSpeed);
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10854

**Comment by joey-wheeler99** [comment]
Is this the same?

```gdscript
#	NOTE: example above
	velocity = Vector3.FORWARD * random_speed
	velocity = velocity.rotated(Vector3.UP, rotation.y)
#	NOTE: mine
	velocity = transform.basis.z * -random_speed
```

**Comment by Dj-Bubbles3** [comment]
so im new to coding and i dont get like anything after the part where you add the move_and_slide() part, like i dont understand how any of the stuff after that works. what is a vector? why does queue_free() kill things? what do the "." 's signify in the code when coming after another thing? what does @export do? what determines what the start_posistion and player_posistion 's values are? is the player_posistion value only called that because of the name of our player being called "player" in the file names, or is it just a thing that is automatically checked by the engine?

idk, im very worried about getting stuck in tutorial purgatory and i have a hard time absorbing the notes about these things. are these dumb questions? idk, im finally trying to make the game after years of writing the story and things aren't looking to good so far :|

**Reply by Hesamom** [reply]
Hey, it's fine to get confused and overwhelmed if it's your first time programming a game, and based on your questitons it seems you're learning too many things at once, I recommend first starting by learning either GDscript or c# without developing a game to understand their syntaxes and concepts.

Afterwards you can review basic linear algebra math to undertand how vectors and 3D transformaions works, if it's too much you can always fallback to 2D first and then get back to 3D since it's easier to begin with.

**Reply by ken-bkz** [reply]
I'm also brand new to coding (....unless you count HTML like twenty years ago)

Harvard's CS50 Programming with Python is free on YouTube. I started with Godot and got overwhelmed. My frustrations were the same as yours... I have no interest in learning Python, but I'll tell you how the course helped me.

Newbies like us need to get exposed to the programming mindsets everyone else seems to *just magically have*. This Harvard course does an amazing job walking you through how to solve problems with code (and not just 'hey copy me, ok got it? lol bye')

Plus, Python is a general use language, meaning it's useful for a huge broad list of *everything*. I'm not suggesting you master it, or even learn it... but if you watch their free course and understand the basic principles then you'll have a much easier time with GDScript because GDScript is specialized.

Or another way to say it: Python is used for all kinds of different programming, GDScript is *just* games. If you understand how Python works then your understanding of GDScript will go up ten-fold.

**Reply by Hakkesaelger** [reply]
did you follow all the previous tutorials? I'd recommend learning a little programming here: https://gdquest.github.io/learn-gdscript/?ref=godot-docs , and then following everything that comes before this in the getting started section of the website, starting at https://docs.godotengine.org/en/4.4/getting_started/introduction/index.html before making complex 3d games like this one

**Comment by Hakkesaelger** [comment]
I did the whole visibleOnScreen thing, and connected the signal, and wrote the line to do the queue_free, but the mobs still aren't disappearing when they leave (which I can see by pressing the "override the camera" button. why is it not working?

**Reply by Hakkesaelger** [reply]
i removed the code and reentered it and it suddenly worked so...

**Comment by Embersember** [comment]
I was wanting to alter the mob spawning direction manually, and not have them chase the player, so I could do some basic patterns, but I can't figure out how to make that happen. I tried to substitute Vector3.RIGHT for player_position in the Mob and Main scenes, but that seemed to change nothing, and I tried to use the format of the 2D mob spawning but that just spawned everything in under the player at an angle and not moving. Here's what I did, if it at all helps, but I think I just tied myself in knots:
func _on_mob_timer_timeout():
	var mob = mob_scene.instantiate()
	
	var mob_spawn_location = get_node("SpawnPath/SpawnLocation")
	mob_spawn_location.progress_ratio = randf()
	
	var direction = mob_spawn_location.rotation + Vector3.DOWN
	mob.rotation = direction
	
	var velocity = Vector3.FORWARD * 10
	velocity = velocity.rotated(Vector3.UP, PI/2)
	
	#var player_position = $Player.position
	#mob.initialize(mob_spawn_location.position, player_position)
	
	add_child(mob)

My main goal with these things was to just make the mobs come out perpendicular to the screen, but I'd also like to be able to specify where they spawn too. That part I think I could do by replacing randf() with specifics, but if there's a better approach to any of this, I'd love to know!

**Comment by admin** [comment]
*Originally posted by [**Mr-Potato-l**](https://github.com/Mr-Potato-l) on 2026-05-28 15:40:22 UTC*

> Hello!
> On this line in the `Initialize()` function:
> `LookAtFromPosition(startPosition, playerPosition, Vector3.Up);'
> If a monster spawns when the player is in the air, the monster then spawns tilted and part of it clips through the ground.

**Comment by sen507451-sys** [comment]
extends CharacterBody3D

# How fast the player moves in meters per second.
@export var speed = 14
# The downward acceleration when in the air, in meters per second squared.
@export var fall_acceleration = 75

var target_velocity = Vector3.ZERO


func _physics_process(delta):
	var direction = Vector3.ZERO

	if Input.is_action_pressed("move_right"):
		direction.x += 1
	if Input.is_action_pressed("move_left"):
		direction.x -= 1
	if Input.is_action_pressed("move_back"):
		direction.z += 1
	if Input.is_action_pressed("move_forward"):
		direction.z -= 1

	if direction != Vector3.ZERO:
		direction = direction.normalized()
		# Setting the basis property will affect the rotation of the node.
		$Pivot.basis = Basis.looking_at(direction)

	# Ground Velocity
	target_velocity.x = direction.x * speed
	target_velocity.z = direction.z * speed

	# Vertical Velocity
	if not is_on_floor(): # If in the air, fall towards the floor. Literally gravity
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)

	# Moving the Character
	velocity = target_velocity
	move_and_slide()
